napari/layers/base/base.py

```diff
@@ -1326,38 +1245,41 @@ def _world_to_displayed_data_ray(
         vector_data_ndisplay /= np.linalg.norm(vector_data_ndisplay)
         return vector_data_ndisplay
 
-    def _world_to_data_dims_displayed(
-        self, dims_displayed: List[int], ndim_world: int
+    def _world_to_layer_dims(
+        self, *, world_dims: List[int], ndim_world: int
     ) -> List[int]:
-        """Convert indices of displayed dims from world to data coordinates.
+        """Map world dimensions to layer dimensions while maintaining order.
 
-        This accounts for differences in dimensionality between the world
-        and the data coordinates. For example a world dims order of
-        [2, 1, 0, 3] would be [0, 1] for a layer that only has two dimensions
-        or [1, 0, 2] for a layer with three as that corresponds to the
-        relative order of the last two and three dimensions respectively
+        This is used to map dimensions from the full world space defined by ``Dims``
+        to the subspace that a layer inhabits, so that those can be used to index the
+        layer's data and associated coordinates.
+
+        For example a world ``Dims.order`` of [2, 1, 0, 3] would map to [0, 1] for a
+        layer with two dimensions and [1, 0, 2] for a layer with three dimensions
+        as those correspond to the relative order of the last two and three world dimensions
+        respectively.
```

Comment:
EDIT: the more I think about this, and the less I feel I understand it... I leave my comment below, but I'm no longer convinced I got it :/

---

Ok, so I got quite confused and double backed a couple times when reading this docstring. I think I understand now, so I propose to add a clearer schematic to the dosctring for future reference. What I understand from the above is basically this (which is WRONG):

```
[2, 1, 0, 3]
    |  |  |
    | [0, 1]
    |  |  |
   [1, 0, 2]
```

However, this makes no sense. Instead, this function returns the *indices* that I can use to index into the layer data corresponding to the world dimensions.

So the right way to look at this is as follows: we right-align dimensions like this:

```
[0, 1, 2, 3]
    |  |  |
    | [0, 1]
    |  |  |
   [0, 1, 2]
```

Then we change the order:

```
[2, 1, 0, 3]
 |  |     |
[0, |     1]
 |  |     |
[1, 0,    2]
```

And we get the final pairings. Sorry if this was obvious to you, but it really confused me :P

What I'm *still* confused about it the role of `ndim_world`. Isn't that just `len(world_dims)`?

Reply:
Don't worry. I'm still confused too. I tried to improve things with some renaming and docstrings, but looks like it didn't help too much :D

I think it's a good goal to of this PR to clarify what is going on now, though I think fixing and simplify everything is out of scope.

Agreed that some better schematics here help. I usually add names to the dimensions of each layer (e.g. `a0`, `b2`), so that I can keep track of the correspondences while being too lazy to maintain ascii art. I'll this though.

As for `ndim_world`, I'm pretty sure it is when this is called in `Layer._slice_dims`, but I'm not sure if that's also true for `_points_mouse_bindings._update_drag_vectors_from_event`. Will check and see if we can simplify that.

Reply:
I think we're basically repeating the conversation in #3882.

My understanding is that the interpretation you got from the current docstring is correct in the sense that this is what napari does now. As @jni put it

> the mapping of dimensions between different arrays changes depending on the dims being displayed!!!

Another way to think about it is that napari is trying to show as much of each layer data as possible.  I added a test to show this.

As discussed in #3882, what napari does now may not be what we want going forward. What you described (right aligning then reordering) is one of the suggestions. One problem with that approach is that the order of `[2, 1, 0, 3]` and `ndisplay` of 2 implies that we're visualizing dimensions 0 and 3 of the first layer, but 0 doesn't have a corresponding dimension in the other two layers, which means that these will be 1D slices and we're unsure if we should show nothing or that row broadcasted.

This PR is definitely not the place to change the current behavior. But I have added the test above (so we know when it changes).

Reply:
Ok, I think I got it now, thanks! I guess the example above it just *happened* to work out the same way with the two different approaches ^^'